Unmanaged-cluster: Protect cli against extra arguments

marckhouzam opened the pull request:
## What this PR does / why we need it

Protects against unexpected behaviour of the unmanaged-cluster cli when too many arguments are specified.

## Which issue(s) this PR fixes

Fixes #4842

## Describe testing done for PR

I called unmanaged-cluster sub-command with zero, one and two arguments and verified arguments were properly checked.
For the `uc create` command, I also checked each combo with and without the `-f` and `-e` flags.
Below is a script that can be sourced that tests all combinations:
```
for cmd in config delete start stop;do
  echo \$ uc $cmd
  uc $cmd 2>&1 |head -3
  echo "============"
  echo \$ uc $cmd c1
  uc $cmd c1 2>&1 |head -3
  echo "============"
  echo \$ uc $cmd c1 c2
  uc $cmd c1 c2 2>&1 |head -3
  echo "============"
done

echo \$ uc list
uc list
echo "============"
echo \$ uc list extra_arg
uc list extra_arg 2>&1 |head -3
echo "============"

echo \$ uc create
uc create 2>&1 |head -3
echo "============"
echo \$ uc create c1 -p p
uc create c1 -p p
echo "============"
echo \$ uc create c1 c2
uc create c1 c2 2>&1 |head -3
echo "============"
echo \$ uc create -f test
uc create -f test
echo "============"
echo \$ uc create -e test
uc create -e test
echo "============"
echo \$ uc create c1 -f test
uc create c1 -f test
echo "============"
echo \$ uc create c1 -e test
uc create c1 -e test
echo "============"
echo \$ uc create c1 c2 -f test
uc create c1 c2 -f test 2>&1 |head -3
echo "============"
echo \$ uc create c1 c2 -e test
uc create c1 c2 -e test 2>&1 |head -3
echo "============"
```

## Special notes for your reviewer

Before `tanzu uc list extra` would work but now it won't.  This is technically not backwards-compatible.

marckhouzam committed:
Argument checking for cmds requiring cluster name

The commit also standardize argument check for the create cmd

Fixes #4842

Signed-off-by: Marc Khouzam <kmarc@vmware.com>

stmcginnis reviewed (dismissed):
> I like the custom message which provides useful info in this case.

I agree, it is nice being able to provide useful feedback.

Maybe a good feature request for cobra:

```
func MaximumNArgs(n int, tooManyErrorMsg string) PositionalArgs {}
```

jpmcb reviewed (dismissed):
Nice! This looks great 👏🏼

harshanarayana commented:
> Maybe a good feature request for cobra:

Is this not already kind of customisable ?

```golang
var CreateCmd = &cobra.Command{
	Use:   "create <cluster name>",
	Short: "Create an unmanaged cluster",
	Long:  createDesc,
	Run:   create,
	Args: func(cmd *cobra.Command, args []string) error {
		if len(args) == 0 {
			return errors.New("Error: A cluster name is required")
		}
		if len(args) != 1 {
			return errors.New("Error: must specify one cluster name to create a cluster")
		}
		return nil
	},
}
```

marckhouzam commented:
> Is this not already kind of customisable ?
[...]
> Can we not this way for all of the unmanaged-cluster CLI commands and standardize the error messages and validation behaviours ?

Good point @harshanarayana, it would be more elegant to use Cobra's `Args` field to do argument validation. Thanks!  I can update the PR that way.

Can I make a similar change for the `create` command @stmcginnis ?  That way we align all the commands.  If you agree, do you prefer a different PR or can we put everything here?

stmcginnis commented:
> That way we align all the commands

Great idea! I'd love to see this done in a consistent way across commands. Up to you whether you want to add it to this PR or do another - I'm good either way.

marckhouzam commented:
I've updated the PR taking all recommendations and suggestions into account.
I've updated the description of the PR to match, which includes a little script that tests all the different combinations of modified commands and arguments.

Finally, there is a potential backwards-compatibility issue as `tanzu uc list extra` used to pass but now it will reject the useless `extra` argument.  I'm not familiar with TCE's backwards-compatibility policy, so if someone can let me know if I should revert that part of the change, I would appreciate.

jpmcb reviewed (dismissed):
> Finally, there is a potential backwards-compatibility issue as tanzu uc list extra used to pass but now it will reject the useless extra argument

That's fine. It was invalid behavior before so therefore a bug which wouldn't be supported in older versions.

This looks good to me!

jpmcb commented:
@marckhouzam - can you squash to one commit?

harshanarayana reviewed (commented):
This one looks great now. Thanks @marckhouzam

marckhouzam commented:
Squashed.
Thanks for the quick review.

stmcginnis reviewed (approved):
Looks great to me.
